agent/immortalstreams/stream.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	// Update state
+	s.mu.Lock()
+	s.lastConnectionAt = time.Now()
+	s.mu.Unlock()
+
+	s.logger.Debug(context.Background(), "client reconnection successful",
+		slog.F("read_seq_num", readSeqNum))
+	return nil
+}
+
+// Close closes the stream
+func (s *Stream) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.closed {
+		return nil
+	}
+
+	s.closed = true
+
+	// Cancel will interrupt any pending BackedPipe operations
+	if s.cancel != nil {
+		s.cancel()
+	}
+
+	// Signal shutdown to any pending reconnect attempts and listeners
+	// Closing the channel wakes all waiters exactly once
+	close(s.shutdownChan)
+
+	// No reconnection waiters in the simplified model.
+
+	// Get references to resources we need to close, but close them outside the mutex
+	// to avoid deadlocks with reconnection attempts
+	pipe := s.pipe
+	localConn := s.localConn
+
+	// Release the mutex before closing resources to avoid deadlocks
+	s.mu.Unlock()
+
+	// Close the backed pipe (this can trigger reconnection attempts, so must be outside mutex)
+	if pipe != nil {
+		if err := pipe.Close(); err != nil {
+			s.logger.Warn(context.Background(), "failed to close backed pipe", slog.Error(err))
+		}
+	}
+
+	// Close connections
+	if localConn != nil {
+		if err := localConn.Close(); err != nil {
+			s.logger.Warn(context.Background(), "failed to close local connection", slog.Error(err))
+		}
+	}
+
+	// Wait for goroutines to finish
+	s.goroutines.Wait()
+
+	// Re-acquire mutex to balance the deferred unlock
+	s.mu.Lock()
+	return nil
+}
+
+// IsConnected returns whether the stream has an active client connection
+func (s *Stream) IsConnected() bool {
+	p := s.pipe
+	if p == nil {
+		return false
+	}
+	return p.Connected()
+}
+
+// LastDisconnectionAt returns when the stream was last disconnected
+func (s *Stream) LastDisconnectionAt() time.Time {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.lastDisconnectionAt
+}
+
+// ToAPI converts the stream to an API representation
+func (s *Stream) ToAPI() codersdk.ImmortalStream {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	stream := codersdk.ImmortalStream{
+		ID:               s.id,
+		Name:             s.name,
+		TCPPort:          s.port,
+		CreatedAt:        s.createdAt,
+		LastConnectionAt: s.lastConnectionAt,
+	}
+
+	if !s.IsConnected() && !s.lastDisconnectionAt.IsZero() {
+		stream.LastDisconnectionAt = &s.lastDisconnectionAt
+	}
+
+	return stream
+}
+
+// GetPipe returns the backed pipe for handling connections
+func (s *Stream) GetPipe() *backedpipe.BackedPipe {
+	return s.pipe
+}
+
+// startCopyingLocked starts the goroutines to copy data from local connection
+// Must be called with mu held
+func (s *Stream) startCopyingLocked() {
+	// Start bidirectional copy using agentssh.Bicopy, but do not close endpoints
+	// on transient disconnects. We tie the lifecycle to the stream's shutdown context;
+	// Bicopy will only return (and close both ends) when the context is canceled.
```

Comment:
Bicopy will also (correctly) return if the `localConn` throws an error. We need that to trigger the stream getting cleaned up, but it looks like the stream persists until it is explicitly deleted or evicted.